### ISSUE TITLE
Parser inconsistent tzlocal() / tzutc()

### ISSUE THREAD

**Issue by leifurhauks** [dateutil/dateutil#349]
It seems that if the current timezone is UTC, parsing a UTC timestamp string such as `"2017-02-23T06:54:00Z"` results in a `datetime` object with `tzinfo=tzlocal()`. Since the timestamp in the string is expclitly in UTC, as opposed to an arbitrary timezone that happens to have the same offset as the local timezone, shouldn't the tzinfo be `tzutc()`?

```ipython
In [1]: import time, os

In [2]: from dateutil import parser

In [3]: time.tzname, time.timezone
Out[3]: (('EST', 'EDT'), 18000)

In [4]: parser.parse("2017-02-23T06:54:00Z")
Out[4]: datetime.datetime(2017, 2, 23, 6, 54, tzinfo=tzutc())

In [5]: os.environ['TZ'] = 'UTC'

In [6]: time.tzset()

In [7]: time.tzname, time.timezone
Out[7]: (('UTC', 'UTC'), 0)

In [8]: parser.parse("2017-02-23T06:54:00Z")
Out[8]: datetime.datetime(2017, 2, 23, 6, 54, tzinfo=tzlocal())
```

**Comment by pganssle** [comment]
This happens because `dateutil` tries to make some context-dependent guesses about what time zone you mean when the time zone is ambiguous. `GMT` could be `Europe/London` or `UTC`, and apparently it prioritizes `local()` over any other interpretation. I personally don't really like this implicit context, but it's what we have.

A somewhat related bug is #320. I think both can be fixed with a "sanity check" when assigning time zones to parsed dates, where, when you are pulling something that's supposed to be a `tzname()` output, you make sure that that `tzname()` *actually* applies at the time of year you've parsed.

Currently I think the best workaround would be to replace "GMT" with "UTC", which is hopefully unambiguous.

**Comment by leifurhauks** [comment]
> This happens because dateutil tries to make some context-dependent guesses about what time zone you mean when the time zone is ambiguous.

That makes sense, in general, but for a timestamp like `"2017-02-23T06:54:00Z"`, the time zone isn't ambiguous, [it's UTC][1]. The resulting `datetime` object really shouldn't be dependent on the system timezone.

Incidentally, it's not clear to me that #356 is a duplicate of this issue; in the snippet pasted there, it looks like the parser is returning a datetime in the wrong timezone altogether. AFAIK, GMT is never equivalent to UTC+01. British Summer Time is, of course, but that's a separate time zone. Some countries, such as Iceland, use GMT year-round and don't observe DST.

[1]: https://en.wikipedia.org/wiki/ISO_8601#UTC

**Comment by pganssle** [comment]
> That makes sense, in general, but for a timestamp like "2017-02-23T06:54:00Z", the time zone isn't ambiguous, it's UTC. The resulting datetime object really shouldn't be dependent on the system timezone

Agreed, I just meant that the reason it's happening is that `dateutil` seems to prefer to use your local zone whenever that is detected. If you change your local zone to `UTC`, `tzutc()` and `tzlocal()` are equivalent, and evidently `dateutil` prefers `tzlocal()`.

> Incidentally, it's not clear to me that #356 is a duplicate of this issue; in the snippet pasted there, it looks like the parser is returning a datetime in the wrong timezone altogether.

I think you're right that it's not a perfect duplicate. I only skimmed this one and merged all three issues in my head. I thought that it was the same underlying issue, but actually I think this one is more a priority issue (explicit call-outs should beat `tzlocal()`, even if the zones are equally appropriate) and the other one is an issue where the wrong zone is attached to a non-ambiguous datetime where datetimes are ambiguous for some portion of the year (e.g. `GMT` during Winter time could be local time or it could be UTC).

Sorry about that. Not sure there's an actual workaround for this one, though luckily it has no actual symptoms.

**Comment by pganssle** [comment]
@leifurhauks I did come up with a workaround for this:

```python
from dateutil import tz
from dateutil.parser import parse, parserinfo

import os
import time


tzinfos = {x: tz.tzutc() for x in parserinfo().UTCZONE}
os.environ['TZ'] = 'UTC'
time.tzset()
print(repr(parse('2014-03-22T01:30Z')))
# datetime.datetime(2014, 3, 22, 1, 30, tzinfo=tzlocal())

print(repr(parse('2014-03-22T01:30Z', tzinfos=tzinfos)))
# datetime.datetime(2014, 3, 22, 1, 30, tzinfo=tzutc())
```

**Comment by toshetah** [comment]
The solution I found for this problem is using isoparser from this package.
Example:
parser.isoparse("2015-01-01T01:01:01Z")
datetime.datetime(2015, 1, 1, 1, 1, 1, tzinfo=tzutc())

**Comment by wjehenddher** [comment]
@toshetah 's workaround is good as long as `dateutil` version is 2.7.0+ https://github.com/dateutil/dateutil/releases/tag/2.7.0

```python 
>>> from dateutil.parser import parse, isoparse
>>> isoparse('2019-02-19T04:10:30Z')
datetime.datetime(2019, 2, 19, 4, 10, 30, tzinfo=tzutc())
>>> parse('2019-02-19T04:10:30Z')
datetime.datetime(2019, 2, 19, 4, 10, 30, tzinfo=tzlocal())
```

**Comment by jlas** [comment]
I'm also encountering this issue, I would expect `"2017-02-23T06:54:00Z"` to be parsed with an explicit utc timezone.

Interestingly the behavior seems to have changed between v2.7.5 and v2.8.1

```python
>>> dateutil.__version__
'2.7.5'
>>> dateutil.parser.parse("2020-04-15T00:00:00Z")
datetime.datetime(2020, 4, 15, 0, 0, tzinfo=tzutc())
```
```python
>>> dateutil.__version__
'2.8.1'
>>> dateutil.parser.parse("2020-04-15T00:00:00Z")
datetime.datetime(2020, 4, 15, 0, 0, tzinfo=tzlocal())
```

**Comment by ffe4** [comment]
This seems to be the same issue as #842 which is on the milestone for the 3.0 release since it is a breaking change. Plans are to release 3.0 in the [near future](https://github.com/dateutil/dateutil/issues/653#issuecomment-607238175).
